pkg/controller/serviceaccount_secret_refresher/serviceaccount_secret_refresher.go

```diff
@@ -0,0 +1,181 @@
+
+func (r *reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	l := r.log.WithField("request", req.String())
+	err := r.reconcile(ctx, l, req)
+	if err != nil {
+		l.WithError(err).Error("Reconciliation failed")
+	}
+	return reconcile.Result{}, err
+}
+
+func (r *reconciler) reconcile(ctx context.Context, l *logrus.Entry, req reconcile.Request) error {
+	if !r.filter(req) {
+		return nil
+	}
+
+	sa := &corev1.ServiceAccount{}
+	if err := r.client.Get(ctx, req.NamespacedName, sa); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("failed to get serviaccount %s: %w", req.String(), err)
+	}
+
+	var imagePullSecetsToKeep []corev1.LocalObjectReference
+	for _, pullSecretRef := range sa.ImagePullSecrets {
+		shouldKeep, err := r.shouldKeepSecret(ctx, req.Namespace, pullSecretRef.Name)
+		if err != nil {
+			return fmt.Errorf("failed to check if docker secret should be kept: %w", err)
+		}
+		if shouldKeep {
+			imagePullSecetsToKeep = append(imagePullSecetsToKeep, pullSecretRef)
+			continue
+		}
+		l.WithField("imagepullsecretname", pullSecretRef.Name).Info("Not keeping image pull secret")
+	}
+
+	var tokenSecretsToKeep []corev1.ObjectReference
+	for _, tokenSecretRef := range sa.Secrets {
+		shouldKeep, err := r.shouldKeepSecret(ctx, req.Namespace, tokenSecretRef.Name)
+		if err != nil {
+			return fmt.Errorf("failed to check if secret should be kept: %w", err)
+		}
+		if shouldKeep {
+			tokenSecretsToKeep = append(tokenSecretsToKeep, tokenSecretRef)
+			continue
+		}
+		l.WithField("tokensecretname", tokenSecretRef.Name).Info("Not keeping token secret")
+	}
+
+	dockerSecretDiffCount, tokenSecretDiffCount := len(imagePullSecetsToKeep)-len(sa.ImagePullSecrets), len(tokenSecretsToKeep)-len(sa.Secrets)
+
+	if dockerSecretDiffCount != 0 || tokenSecretDiffCount != 0 {
+		l.WithField("docker_secret_diff_count", dockerSecretDiffCount).WithField("token_secret_diff_count", tokenSecretDiffCount).Info("Updating ServiceAccount")
+		sa.ImagePullSecrets = imagePullSecetsToKeep
+		sa.Secrets = tokenSecretsToKeep
+		if err := r.client.Update(ctx, sa); err != nil {
+			return fmt.Errorf("failed to update ServiceAccount: %w", err)
+		}
+	}
+
+	if err := wait.Poll(time.Second, 30*time.Second, func() (bool, error) {
+		if err := r.client.Get(ctx, req.NamespacedName, sa); err != nil {
+			return false, err
+		}
+		// Secrets includes the pull secret, hence the >= 2 instead of >=1
+		return len(sa.ImagePullSecrets) >= 1 && len(sa.Secrets) >= 2, nil
+	}); err != nil {
+		return fmt.Errorf("failed to wait for ServiceAccount to have a pull and a tokenSecret: %w", err)
+	}
+
+	if !r.removeOldSecrets {
+		return nil
+	}
+
+	var allSecrets corev1.SecretList
+	if err := r.client.List(ctx, &allSecrets, ctrlruntimeclient.InNamespace(req.Namespace)); err != nil {
+		return fmt.Errorf("failed to list secrets in %s namespace: %w", req.Namespace, err)
+	}
+
+	for _, secret := range allSecrets.Items {
```

Comment:
What's the trigger for this reconciliation if no SA/Secrets change?

Reply:
the fact that we restart the controller at least daily

Reply:
I'll add a reconcile after

Reply:
fixed